From f4435491dd0492da85b8c98ece46813ad1acc63f Mon Sep 17 00:00:00 2001
From: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Date: Tue, 5 Mar 2024 13:37:26 +0100
Subject: [PATCH 531/841] refactor(st-rif): cleanup RIFSC IP configuration

Move related source code from platform level to RIFSC driver by adding
a dedicated service with adapted definitions.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 15f6ade87d05d6c234e22bf23f8218a63efd2e7f refactor(st-rif): cleanup RIFSC IP configuration]
Signed-off-by: Nicolas Le Bayon <nicolas.le.bayon@st.com>
Change-Id: I000e048ce5085e5e8a6bda82330df1877049cf63
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/364524
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
---
 drivers/st/rif/stm32_rifsc.c            | 58 ++++++++++++++-----------
 include/drivers/st/stm32_rifsc.h        |  3 +-
 include/drivers/st/stm32mp_rifsc_regs.h | 44 ++++++++++++-------
 plat/st/stm32mp2/bl2_plat_setup.c       | 17 ++------
 plat/st/stm32mp2/stm32mp2_def.h         | 16 +++----
 5 files changed, 70 insertions(+), 68 deletions(-)

diff --git a/drivers/st/rif/stm32_rifsc.c b/drivers/st/rif/stm32_rifsc.c
index b0430bb09..04dc0fbb0 100644
--- a/drivers/st/rif/stm32_rifsc.c
+++ b/drivers/st/rif/stm32_rifsc.c
@@ -1,33 +1,21 @@
 /*
- * Copyright (c) 2023, STMicroelectronics - All Rights Reserved
+ * Copyright (c) 2023-2024, STMicroelectronics - All Rights Reserved
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
 
+#include <assert.h>
 #include <errno.h>
 #include <stdint.h>
 
 #include <arch_helpers.h>
 #include <drivers/st/stm32_rifsc.h>
+#include <drivers/st/stm32mp_rifsc_regs.h>
 #include <dt-bindings/soc/rif.h>
 #include <lib/mmio.h>
 
 #include <platform_def.h>
 
-/* RIFSC general register field description */
-/* RIFSC_CIDCFGR register fields */
-#define _RIFSC_RISC_CIDCFGR(x)		(U(0x100) + U(0x8) * (x))
-#define _RIFSC_RISC_CFEN_MASK		BIT(0)
-#define _RIFSC_RISC_SEM_EN_MASK		BIT(1)
-#define _RIFSC_RISC_SEML_SHIFT		U(16)
-#define _RIFSC_RISC_SEML_MASK		GENMASK_32(23, 16)
-
-/* RIFSC_SEMCR register fields */
-#define _RIFSC_RISC_SEMCR(x)		(U(0x104) + U(0x8) * (x))
-#define _RIFSC_RISC_SEM_MUTEX		BIT(0)
-#define _RIFSC_RISC_SEMCID_SHIFT	U(4)
-#define _RIFSC_RISC_SEMCID_MASK		GENMASK_32(6, 4)
-
 static unsigned long rifsc_semaphores[] = {
 	STM32MP25_RIFSC_RNG_ID,
 	STM32MP25_RIFSC_PKA_ID,
@@ -40,29 +28,47 @@ int stm32_rifsc_semaphore_init(void)
 
 	for (i = 0; i < ARRAY_SIZE(rifsc_semaphores); i++) {
 		uint32_t cidcfgr = mmio_read_32(RIFSC_BASE +
-						_RIFSC_RISC_CIDCFGR(rifsc_semaphores[i]));
+						_RIFSC_RISC_PERy_CIDCFGR(rifsc_semaphores[i]));
 		uint32_t semcfgr = mmio_read_32(RIFSC_BASE +
-						_RIFSC_RISC_SEMCR(rifsc_semaphores[i]));
+						_RIFSC_RISC_PERy_SEMCR(rifsc_semaphores[i]));
+
+		uint32_t sem_wl = (cidcfgr & _RIFSC_CIDCFGR_SEML_MASK) >> _RIFSC_CIDCFGR_SEML_SHIFT;
 
-		if (!(((cidcfgr & _RIFSC_RISC_CFEN_MASK) != 0U) &&
-		      ((cidcfgr & _RIFSC_RISC_SEM_EN_MASK) != 0U) &&
-		      ((cidcfgr & _RIFSC_RISC_SEML_MASK) >> _RIFSC_RISC_SEML_SHIFT) != RIF_CID1)) {
+		if (!(((cidcfgr & _RIFSC_CIDCFGR_CFEN) != 0U) &&
+		      ((cidcfgr & _RIFSC_CIDCFGR_SEM_EN) != 0U) &&
+		      ((sem_wl & RIF_CID1_BF) != RIF_CID1_BF))) {
 			continue;
 		}
 
-		if (((semcfgr & _RIFSC_RISC_SEM_MUTEX) != 0U) &&
-		    ((semcfgr & _RIFSC_RISC_SEMCID_MASK) >> _RIFSC_RISC_SEMCID_SHIFT) != RIF_CID1) {
+		if (((semcfgr & _RIFSC_SEMCR_SEM_MUTEX) != 0U) &&
+		    ((semcfgr & _RIFSC_SEMCR_SEMCID_MASK) >> _RIFSC_SEMCR_SEMCID_SHIFT) != RIF_CID1) {
 			return -EACCES;
 		}
 
-		mmio_write_32(RIFSC_BASE + _RIFSC_RISC_SEMCR(rifsc_semaphores[i]),
-			      _RIFSC_RISC_SEM_MUTEX);
+		mmio_write_32(RIFSC_BASE + _RIFSC_RISC_PERy_SEMCR(rifsc_semaphores[i]),
+			      _RIFSC_SEMCR_SEM_MUTEX);
 
-		if (((semcfgr & _RIFSC_RISC_SEM_MUTEX) != 0U) &&
-		    ((semcfgr & _RIFSC_RISC_SEMCID_MASK) >> _RIFSC_RISC_SEMCID_SHIFT) != RIF_CID1) {
+		if (((semcfgr & _RIFSC_SEMCR_SEM_MUTEX) != 0U) &&
+		    ((semcfgr & _RIFSC_SEMCR_SEMCID_MASK) >> _RIFSC_SEMCR_SEMCID_SHIFT) != RIF_CID1) {
 			return -EACCES;
 		}
 	}
 
 	return 0;
 }
+
+void stm32_rifsc_ip_configure(int rimu_id, int rifsc_id, uint32_t param)
+{
+	uint32_t bit;
+
+	assert(rifsc_id < STM32MP25_RIFSC_MAX_ID);
+
+	bit = BIT(rifsc_id / U(32));
+
+	/* Set peripheral accesses to Secure/Privilege only */
+	mmio_write_32(RIFSC_BASE + _RIFSC_RISC_SECCFGR(rifsc_id), bit);
+	mmio_write_32(RIFSC_BASE + _RIFSC_RISC_PRIVCFGR(rifsc_id), bit);
+
+	/* Apply specific configuration to RIF master */
+	mmio_write_32(RIFSC_BASE + _RIFSC_RIMC_ATTR(rimu_id), param);
+}
diff --git a/include/drivers/st/stm32_rifsc.h b/include/drivers/st/stm32_rifsc.h
index b720eac98..56757a10d 100644
--- a/include/drivers/st/stm32_rifsc.h
+++ b/include/drivers/st/stm32_rifsc.h
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2023, STMicroelectronics - All Rights Reserved
+ * Copyright (c) 2023-2024, STMicroelectronics - All Rights Reserved
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -13,5 +13,6 @@
  * Return 0 on success and -EACCES if authorized semaphore couldn't be taken
  */
 int stm32_rifsc_semaphore_init(void);
+void stm32_rifsc_ip_configure(int rimu_id, int rifsc_id, uint32_t param);
 
 #endif /* STM32MP2_RIFSC_H */
diff --git a/include/drivers/st/stm32mp_rifsc_regs.h b/include/drivers/st/stm32mp_rifsc_regs.h
index b393af34a..5b563faa2 100644
--- a/include/drivers/st/stm32mp_rifsc_regs.h
+++ b/include/drivers/st/stm32mp_rifsc_regs.h
@@ -8,22 +8,22 @@
 #define STM32MP_RIFSC_REGS_H
 
 /* RIFSC offset register */
-#define _RIFSC_RISC_CR			U(0x0)
-#define _RIFSC_RISC_SECCFGR(id)		(U(0x10) + U(0x4) * ((id) / 32))
-#define _RIFSC_RISC_PRIVCFGR(id)	(U(0x30) + U(0x4) * ((id) / 32))
-#define _RIFSC_RISC_RCFGLOCKR(id)	(U(0x50) + U(0x4) * ((id) / 32))
-#define _RIFSC_RISC_PERy_CIDCFGR(id)	(U(0x100) + U(0x8) * (id))
-#define _RIFSC_RISC_PERy_SEMCR(id)	(U(0x104) + U(0x8) * (id))
-#define _RIFSC_RIMC_CR			U(0xC00)
-#define _RIFSC_RIMC_SR			U(0xC04)
-#define _RIFSC_RIMC_ATTR(x)		(U(0xC10) + U(0x4) * (x))
-#define _RIFSC_PPSR(x)			(U(0xFB0) + U(0x4) * (x))
-#define _RIFSC_HWCFGR3			U(0xFE8)
-#define _RIFSC_HWCFGR2			U(0xFEC)
-#define _RIFSC_HWCFGR1			U(0xFF0)
-#define _RIFSC_VERR			U(0xFF4)
-#define _RFISC_IPIDR			U(0xFF8)
-#define _RFISC_SIDR			U(0xFFC)
+#define _RIFSC_RISC_CR				U(0x0)
+#define _RIFSC_RISC_SECCFGR(id)			(U(0x10) + U(0x4) * ((id) / 32))
+#define _RIFSC_RISC_PRIVCFGR(id)		(U(0x30) + U(0x4) * ((id) / 32))
+#define _RIFSC_RISC_RCFGLOCKR(id)		(U(0x50) + U(0x4) * ((id) / 32))
+#define _RIFSC_RISC_PERy_CIDCFGR(id)		(U(0x100) + U(0x8) * (id))
+#define _RIFSC_RISC_PERy_SEMCR(id)		(U(0x104) + U(0x8) * (id))
+#define _RIFSC_RIMC_CR				U(0xC00)
+#define _RIFSC_RIMC_SR				U(0xC04)
+#define _RIFSC_RIMC_ATTR(x)			(U(0xC10) + U(0x4) * (x))
+#define _RIFSC_PPSR(x)				(U(0xFB0) + U(0x4) * (x))
+#define _RIFSC_HWCFGR3				U(0xFE8)
+#define _RIFSC_HWCFGR2				U(0xFEC)
+#define _RIFSC_HWCFGR1				U(0xFF0)
+#define _RIFSC_VERR				U(0xFF4)
+#define _RFISC_IPIDR				U(0xFF8)
+#define _RFISC_SIDR				U(0xFFC)
 
 /* RIFSC_RIMC_ATTRx register fields */
 #define RIFSC_RIMC_ATTRx_CIDSEL			BIT_32(2)
@@ -32,4 +32,16 @@
 #define RIFSC_RIMC_ATTRx_MSEC			BIT_32(8)
 #define RIFSC_RIMC_ATTRx_MPRIV			BIT_32(9)
 
+
+/* RIFSC_RISC_PERy_CIDCFGR register fields */
+#define _RIFSC_CIDCFGR_CFEN			BIT_32(0)
+#define _RIFSC_CIDCFGR_SEM_EN			BIT_32(1)
+#define _RIFSC_CIDCFGR_SEML_SHIFT		U(16)
+#define _RIFSC_CIDCFGR_SEML_MASK		GENMASK_32(23, 16)
+
+/* RIFSC_RISC_PERy_SEMCR register fields */
+#define _RIFSC_SEMCR_SEM_MUTEX			BIT_32(0)
+#define _RIFSC_SEMCR_SEMCID_SHIFT		U(4)
+#define _RIFSC_SEMCR_SEMCID_MASK		GENMASK_32(6, 4)
+
 #endif /* STM32MP_RIFSC_REGS_H */
diff --git a/plat/st/stm32mp2/bl2_plat_setup.c b/plat/st/stm32mp2/bl2_plat_setup.c
index 198124576..2edce1c38 100644
--- a/plat/st/stm32mp2/bl2_plat_setup.c
+++ b/plat/st/stm32mp2/bl2_plat_setup.c
@@ -26,7 +26,6 @@
 #include <drivers/st/stm32_saes.h>
 #include <drivers/st/stm32mp_pmic2.h>
 #include <drivers/st/stm32mp_reset.h>
-#include <drivers/st/stm32mp_rifsc_regs.h>
 #include <drivers/st/stm32mp_risab_regs.h>
 #include <drivers/st/stm32mp2_ddr_helpers.h>
 #include <drivers/st/stm32mp2_ram.h>
@@ -402,20 +401,10 @@ skip_console_init:
 	mmio_write_32(RISAB3_BASE + RISAB_CR, RISAB_CR_SRWIAD);
 #endif
 #endif /* STM32MP_DDR_FIP_IO_STORAGE || TRUSTED_BOARD_BOOT */
-#if STM32MP_USB_PROGRAMMER
-	/*
-	 * Set USB3DR Peripheriphal accesses to Secure/Privilege only
-	 */
-	mmio_write_32(RIFSC_BASE + _RIFSC_RISC_SECCFGR(STM32MP25_RIFSC_USB3DR_ID),
-		     RIFSC_USB3DR_SEC);
-	mmio_write_32(RIFSC_BASE + _RIFSC_RISC_PRIVCFGR(STM32MP25_RIFSC_USB3DR_ID),
-		      RIFSC_USB3DR_PRIV);
 
-	/*
-	 * Apply USB boot specific configuration to RIF master USB3DR
-	 */
-	mmio_write_32(RIFSC_BASE + _RIFSC_RIMC_ATTR(STM32MP25_RIMU_USB3DR),
-		      RIFSC_USB_BOOT_USBDR_RIMC_CONF);
+#if STM32MP_USB_PROGRAMMER
+	stm32_rifsc_ip_configure(STM32MP2_RIMU_USB3DR, STM32MP25_RIFSC_USB3DR_ID,
+				 RIFSC_USB_BOOT_USB3DR_RIMC_CONF);
 #endif /* STM32MP_USB_PROGRAMMER */
 
 #if !STM32MP_M33_TDCID
diff --git a/plat/st/stm32mp2/stm32mp2_def.h b/plat/st/stm32mp2/stm32mp2_def.h
index 863857761..b74fa612b 100644
--- a/plat/st/stm32mp2/stm32mp2_def.h
+++ b/plat/st/stm32mp2/stm32mp2_def.h
@@ -10,6 +10,7 @@
 #include <common/tbbr/tbbr_img_def.h>
 #ifndef __ASSEMBLER__
 #include <drivers/st/bsec.h>
+#include <drivers/st/stm32mp_rifsc_regs.h>
 #include <drivers/st/stm32mp2_clk.h>
 #include <drivers/st/stm32mp2_risaf.h>
 #endif
@@ -35,6 +36,7 @@
 #include <dt-bindings/clock/stm32mp25-clks.h>
 #include <dt-bindings/clock/stm32mp25-clksrc.h>
 #include <dt-bindings/reset/stm32mp25-resets.h>
+#include <dt-bindings/soc/stm32mp25-rif.h>
 #endif /* STM32MP25 */
 #include <dt-bindings/gpio/stm32-gpio.h>
 #include <dt-bindings/soc/rif.h>
@@ -632,22 +634,14 @@ enum ddr_type {
 #define RISAF_SEED_SIZE_IN_BYTES		U(4)
 
 /*******************************************************************************
- * RIFSC IDs
+ * RIFSC
  ******************************************************************************/
-#define STM32MP25_RIFSC_USB3DR_ID	U(66)
-#define STM32MP25_RIFSC_RNG_ID		U(92)
-#define STM32MP25_RIFSC_PKA_ID		U(93)
-#define STM32MP25_RIFSC_SAES_ID		U(94)
-
-#define STM32MP25_RIMU_USB3DR		U(4)
-
-#define RIFSC_USB3DR_PRIV		BIT(STM32MP25_RIFSC_USB3DR_ID / U(32))
-#define RIFSC_USB3DR_SEC		BIT(STM32MP25_RIFSC_USB3DR_ID / U(32))
+#define STM32MP2_RIMU_USB3DR		U(4)
 
 /*
  * USB3DR Secure/Priv Master (DMA) access
  */
-#define RIFSC_USB_BOOT_USBDR_RIMC_CONF	(RIFSC_RIMC_ATTRx_MPRIV | RIFSC_RIMC_ATTRx_MSEC | \
+#define RIFSC_USB_BOOT_USB3DR_RIMC_CONF	(RIFSC_RIMC_ATTRx_MPRIV | RIFSC_RIMC_ATTRx_MSEC | \
 					 RIF_CID1 << RIFSC_RIMC_ATTRx_MCID_SHIFT | \
 					 RIFSC_RIMC_ATTRx_CIDSEL)
 
-- 
2.47.2

